File: 01-Languages/Go/13-Generics/main.go
// main.go - generic functions, type constraints, generic types.
package main

import "fmt"

func First[T any](items []T) T {
	return items[0]
}

type Number interface {
	int | int64 | float64
}

func Sum[T Number](items []T) T {
	var total T
	for _, item := range items {
		total += item
	}
	return total
}

type Stack[T any] struct {
	items []T
}

func (s *Stack[T]) Push(item T) {
	s.items = append(s.items, item)
}

func (s *Stack[T]) Pop() T {
	n := len(s.items)
	item := s.items[n-1]
	s.items = s.items[:n-1]
	return item
}

func (s *Stack[T]) Len() int {
	return len(s.items)
}

func main() {
	fmt.Println("--- generic function with inference ---")
	fmt.Println(First([]int{1, 2, 3}))
	fmt.Println(First([]string{"a", "b"}))

	fmt.Println("\n--- constrained generic (Number) ---")
	fmt.Println("Sum of ints:", Sum([]int{1, 2, 3, 4}))
	fmt.Println("Sum of floats:", Sum([]float64{1.5, 2.5}))

	fmt.Println("\n--- generic type Stack[T] ---")
	numberStack := &Stack[int]{}
	numberStack.Push(1)
	numberStack.Push(2)
	numberStack.Push(3)
	fmt.Println("numberStack.Len():", numberStack.Len())
	fmt.Println("numberStack.Pop():", numberStack.Pop())

	stringStack := &Stack[string]{}
	stringStack.Push("a")
	stringStack.Push("b")
	fmt.Println("stringStack.Pop():", stringStack.Pop())
}

File: 01-Languages/Rust/13-Generics/main.rs
// main.rs - generic functions, trait bounds, generic structs.

fn first<T>(items: &[T]) -> &T {
    &items[0]
}

fn largest<T: PartialOrd + Copy>(items: &[T]) -> T {
    let mut largest = items[0];
    for &item in items {
        if item > largest {
            largest = item;
        }
    }
    largest
}

struct Stack<T> {
    items: Vec<T>,
}

impl<T> Stack<T> {
    fn new() -> Self {
        Stack { items: Vec::new() }
    }
    fn push(&mut self, item: T) {
        self.items.push(item);
    }
    fn pop(&mut self) -> Option<T> {
        self.items.pop()
    }
    fn len(&self) -> usize {
        self.items.len()
    }
}

fn main() {
    println!("--- generic function with inference ---");
    println!("{}", first(&[1, 2, 3]));
    println!("{}", first(&["a", "b"]));

    println!("\n--- trait-bounded generic function ---");
    println!("largest of ints: {}", largest(&[3, 7, 2, 9, 4]));
    println!("largest of floats: {}", largest(&[1.5, 9.9, 2.2]));

    println!("\n--- generic struct Stack<T> ---");
    let mut number_stack: Stack<i32> = Stack::new();
    number_stack.push(1);
    number_stack.push(2);
    number_stack.push(3);
    println!("number_stack.len(): {}", number_stack.len());
    println!("number_stack.pop(): {:?}", number_stack.pop());

    let mut string_stack: Stack<String> = Stack::new();
    string_stack.push("a".to_string());
    string_stack.push("b".to_string());
    println!("string_stack.pop(): {:?}", string_stack.pop());

    let mut empty_stack: Stack<i32> = Stack::new();
    println!("empty_stack.pop(): {:?}", empty_stack.pop());
}
